### DISCUSSION TITLE
Adding CSS Attribute value pairs and more to completion data

### DISCUSSION THREAD

**Discussion by AnthonyMuscio** [Gk0Wk/TW5-CodeMirror-Enhanced#26] [Ideas]
Hi,

I was led here from https://talk.tiddlywiki.org/t/looking-for-css-attributes-and-values-resource-data/2037?u=tw_tones there I was asking for souces of Looking for **CSS attributes and Values resource (data).** If anyone in this project can help me please join that conversation.

In the meantime I thought I would mention it here so you can see the potential. If we can take structured information and embed it in a data tiddler or plugin such that the completion process uses this to match keywords and make suggestions, including subsequent values we would open tiddlywiki through the codemirror plugin to support training and development activities in any domain for which we can compile the structured data eg;

- CSS Attribute value pairs
- HTML elements and attributes 
- Other programing language syntax
- Various batch programing languages batch/powershell etc...
- Domain specific data like geological epochs or Biological trees
- Custom structures that develop as content is added to a tiddlywiki

Of course when we stray from plan language terms and keywords, into symbolic code it may be more complex but it seems smart to start with words or pseudowords eg CSS

background
- background-attachment
  - _values_
- background-blend-mode
  - values
- background-clip
  - values
- background-color
  - values
- - ...

**Comment by linonetwo** [comment]
I know, you mean adding Language server protocol to the completion, and adding LSP plugins to TiddlyWiki plugin library. (LSP is the way VSCode and Vim support completion.)

But currently, async completion is not possible in code mirror, and how to adapt LSP as tiddlywiki  plugin is still unknown.
